Desktop, Mobile: Link resource Icon

CalebJohn opened the pull request:
This pull customizes the resource link icon based on the mimeType of the attached item. A screenshot will demonstrate what it does best
![image](https://user-images.githubusercontent.com/2179547/67714928-81ced980-f98e-11e9-86e7-40e23469d456.png)

The icon will follow the current set theme, additionally the icons can be set using css by overwriting the fork-awesome css.
As you can see this only applies to resources so regular urls won't have an icon.

unfortunately I had a lot of trouble trying to get this working on mobile, which is why for now I have it disabled on mobile. @laurent22 if that doesn't work, could you help me out in getting mobile working? I tried copying the way that katex loads fonts at runtime, but I could never get the icons to render properly, even with that.

Based on [the forum post about this](https://discourse.joplinapp.org/t/set-of-icons-for-attachments/2813) maybe we should discuss exactly how we want resource/link icons to look before merging this.

laurent22 commented:
Thanks @CalebJohn, and the code looks good. I think we can indeed get it working on mobile so I'll take a look.

laurent22 commented:
Ok it's a bit complicated to get this working on mobile due to the font files that need to be embedded, so let's focus on desktop only for now.

I think the icons look a bit big, could you make them a bit smaller? In my tests 1.4em looked good. Also if you could add a bit more margin to the right of the icon (0.4em looked fine for me).

Finally, if I attach a .txt file, there's no icon for it, while I expect it's a commonly used file. Since we use Font Awesome lib to get the icon name, maybe it's a different name in Fork Awesome?

CalebJohn commented:
I've changed them as you suggested, it's funny, I originally had 1.4em and 0.4em, but as I tweaked things around I settled on slightly different values. 

You were absolutely right about the txt issue, turns out that font awesome doesn't have a txt format, I added a special case for it and checked the rest of the icons to make ensure they're present in both.

In regards to the mobile issue, we can actually change the way we render icons back to the old method (inline svg) that tessus had added (which is still the fallback for mobile). I would only have to add ~11 inline svg icons. And we could use the font awesome names to clean up the code a bit!
What do you think? If it's alright with you I might add a commit here that would replace the font stuff.

laurent22 commented:
> In regards to the mobile issue, we can actually change the way we render icons back to the old method (inline svg) that tessus had added (which is still the fallback for mobile). I would only have to add ~11 inline svg icons. And we could use the font awesome names to clean up the code a bit!
> What do you think? If it's alright with you I might add a commit here that would replace the font stuff.

That sounds like a good solution and if you could figure out how to implement it that would be great. I guess we would only support the 11 most common file formats then, but that's probably more than enough.

CalebJohn commented:
Fork awesome only really has that many so it shouldn't be an issue, additionally the library that we use (font-awesome-filetypes) only supports 12 (one of which is csv which isn't in fork awesome).

CalebJohn commented:
I finished the updates we talked about. Also tested on mobile!

Here is all supported icons together
![image](https://user-images.githubusercontent.com/2179547/68062148-2b370780-fcce-11e9-8e26-3c37dd2c3416.png)

laurent22 commented on ReactNativeClient/lib/renderers/noteStyle.js:
```diff
@@ -105,10 +105,45 @@ module.exports = function(style, options) {
+		}
+		.fa-file {
+			-webkit-mask: url("data:image/svg+xml;utf8,<svg viewBox='0 0 1536 1992' xmlns='http://www.w3.org/2000/svg'><path d='M1468 380c37 37 68 111 68 164v1152c0 53-43 96-96 96H96c-53 0-96-43-96-96V96C0 43 43 0 96 0h896c53 0 127 31 164 68zm-444-244v376h376c-6-17-15-34-22-41l-313-313c-7-7-24-16-41-22zm384 1528V640H992c-53 0-96-43-96-96V128H128v1536h1280z'/></svg>");
+		}
```
Could you please add a comment to clarify where these fonts come from and how to update them (or add new ones)?

CalebJohn replied:
Done!

laurent22 commented:
Thanks @CalebJohn, I've just left one comment and once this is done we can merge.

tessus commented:
If there's zip, there should be the other compressed formats too:

.tar.gz
.tar.xz
.7z

Just my 2 cents. Or are these included in `file-archive`?

CalebJohn commented:
@tessus tar.gz does work, but it looks like the other 2 don't. I was thinking I would probably make a pull adding some filetypes because right now what they have is very limited.

tessus commented:
@CalebJohn it's not necessary for this PR. I just thought that there are way more compressed formats than just zip and it would have been nice to recognize those as well.

laurent22 commented:
It's all good now, many thanks for the PR @CalebJohn! For the file types we can indeed support more afterwards. I would assume they already map things like video and image mime types to the correct icons, but perhaps some are missing.
